lib/ABI/ModelAlignment.cpp

```diff
@@ -0,0 +1,171 @@
+  }
+
+  rc_return Alignment;
+}
+
+static RecursiveCoroutine<std::optional<uint64_t>>
+naturalAlignmentImpl(model::VerifyHelper &VH,
+                     const model::Type &Type,
+                     model::ABI::Values ABI) {
+  auto MaybeAlignment = VH.size(&Type);
+  if (MaybeAlignment)
+    rc_return MaybeAlignment;
+
+  // This code assumes that the type `Type` is well formed.
+  uint64_t Alignment;
+
+  switch (Type.Kind) {
+  case model::TypeKind::RawFunctionType:
+  case model::TypeKind::CABIFunctionType:
+    // Function prototypes have no size - hence no alignment.
+    Alignment = 0;
+    break;
+
+  case model::TypeKind::PrimitiveType:
+    // The alignment of primitives is simple to figure out based on the abi
+    if (const auto *P = llvm::cast<model::PrimitiveType>(&Type);
+        P->PrimitiveKind == model::PrimitiveTypeKind::Void) {
+      // Make sure void types have no size.
+      if (P->Size != 0)
+        rc_return std::nullopt;
+
+      Alignment = 0;
+    } else {
+      Alignment = abi::getScalarStackAlignment(ABI)[P->Size].AlignAt;
+    }
+    break;
+
+  case model::TypeKind::EnumType: {
+    // The alignment of an enum is the same as the alignment of its underlying
+    // type
+    const auto &Underlying = llvm::cast<model::EnumType>(&Type)->UnderlyingType;
+    if (auto A = rc_recur naturalAlignmentImpl(VH, Underlying, ABI))
+      Alignment = *A;
+    else
+      rc_return std::nullopt;
+  } break;
+
+  case model::TypeKind::TypedefType: {
+    // The alignment of an enum is the same as the alignment of its underlying
+    // type
+    const auto &Typedef = llvm::cast<model::TypedefType>(&Type)->UnderlyingType;
+    if (auto A = rc_recur naturalAlignmentImpl(VH, Typedef, ABI))
+      Alignment = *A;
+    else
+      rc_return std::nullopt;
+  } break;
+
+  case model::TypeKind::StructType: {
+    // The alignment of a struct is the same as the alignment of its most
+    // strictly aligned member.
+    const auto &Fields = llvm::cast<model::StructType>(&Type)->Fields;
+    if (auto A = rc_recur aggregateAlignmentImpl(VH, Fields, ABI))
+      Alignment = *A;
+    else
+      rc_return std::nullopt;
+  } break;
+
+  case model::TypeKind::UnionType: {
+    // The alignment of a union is the same as the alignment of its most
+    // strictly aligned member.
+    const auto &Fields = llvm::cast<model::UnionType>(&Type)->Fields;
+    if (auto A = rc_recur aggregateAlignmentImpl(VH, Fields, ABI))
+      Alignment = *A;
+    else
+      rc_return std::nullopt;
+  } break;
+
+  case model::TypeKind::Invalid:
+  case model::TypeKind::Count:
+  default:
+    rc_return std::nullopt;
+  }
+
+  VH.setSize(&Type, Alignment);
+
+  rc_return Alignment;
+}
+
+static RecursiveCoroutine<std::optional<uint64_t>>
+naturalAlignmentImpl(model::VerifyHelper &VH,
+                     const model::QualifiedType &Type,
+                     model::ABI::Values ABI) {
+  // This code assumes that the QualifiedType is well formed.
```

Comment:
No `.` at end of comments.

Reply:
`grep`ping the revng codebase, there are currently 2320 comments that end with a `.`, and 5269 comments that do not (a lot of which are also multi-line comments which wouldn't need one anyway) which seems kind of inconsistent. We should probably standardize it sooner or later and enforce it using the `check-conventions`. What do you think?

Reply:
The convention is that single statement comments have no trailing `.`. Not easy to enforce automatically `:(`
```
git grep '// [A-Z].*\.$' | grep -v 'file is distributed under the MIT License. See LICENSE.md for '
```
Usually that's a sign you're the author of the code.

Reply:
Whoops 😅.

What can I do, sentences with `.` at the end just look neater for me :0

Reply:
(I'll keep this one open for now as well so that I can take care of any additional `.`s that might emerge during the review)